File: Practica01/code/CountingSort/CountingSort.py
from sys import stdin
import time

def countSort(a):
    size = len(a)
    output = (size+1) *[0]
    maximo1 = max(a)
    count = (maximo1+1)*[0]

    for i in range(0,maximo1+1):
        count[i] = 0

    for i in range(0,size):
        count[a[i]]= count[a[i]] + 1

    for i in range(1,maximo1+1):
        count[i] += count[i-1]

    for i in range(size-1,-1,-1):
        output[count[a[i]]-1] = a[i]
        count[a[i]] -= 1

    for i in range(0,size):
        a[i] = output[i]

if __name__ == '__main__':
    n = stdin.readline()
    while (n):
        arr = [int(x) for x in stdin.readline().split()]
        start = time.time()
        countSort(arr)
        end = time.time()
        #print(arr)
        print(round(end-start, 4), end=' ')
        n = stdin.readline()
    print()

File: Practica01/code/CountingSort/CountingSort.cpp
#include <iostream>
#include <vector>
#include <chrono>

using namespace std;

int getMax(int a[],int b)
{
   int max = a[0];
   for(int i = 1; i<b; i++)
   {
      if(a[i] > max) max = a[i];
   }
   return max;
}

void countSort(int a[],int b)
{
    int size = b;
    int output[size];
    int max = getMax(a,size);
    int count[max+1];

    for(int i = 0; i<max+1; i++) count[i] = 0;
    for(int i = 0; i <size; i++) count[a[i]]++;
    for(int i = 1; i<max+1; i++) count[i] += count[i-1];

    for(int i = size-1; i>=0; i--)
    {
        output[count[a[i]]-1] = a[i];
        count[a[i]] -= 1;
    }

    for(int i = 0; i<size; i++)
    {
        a[i] = output[i];
    }
}

void printArray(int a[], int n) {
  for (int i = 0; i < n; ++i) {
    printf("%d ", a[i]);
  }
  printf("\n");
}

int main() {
  int n = 0;
  while (std::cin >> n) {
    int arr[n] = {0};
    for (int i = 0; i < n; ++i) {
      std::cin >> *(arr + i);
    }
    auto start = std::chrono::steady_clock::now();
    countSort(arr, n);
    auto end = std::chrono::steady_clock::now();
    // printf("Sorted array: \n");
    // printArray(arr, n);
    std::chrono::duration<double> elapsed_seconds = end - start;
    std::cout << elapsed_seconds.count() << ' ';
  }
  std::cout << '\n';

  return 0;
}
